server/core/src/main/java/dev/slimevr/tracking/processor/Constraint.kt

```diff
@@ -0,0 +1,183 @@
+package dev.slimevr.tracking.processor
+
+import io.github.axisangles.ktmath.Quaternion
+import io.github.axisangles.ktmath.Vector3
+import kotlin.math.*
+
+/**
+ * Represents a function that applies a rotational constraint.
+ */
+typealias ConstraintFunction = (inputRotation: Quaternion, thisBone: Bone, limit1: Float, limit2: Float, limit3: Float) -> Quaternion
+
+/**
+ * Represents the rotational limits of a Bone relative to its parent,
+ * twist and swing are the max and min when constraintType is a hinge.
+ */
+class Constraint(
+	val constraintType: ConstraintType,
+	twist: Float = 0.0f,
+	swing: Float = 0.0f,
+	allowedDeviation: Float = 0f,
+	maxDeviationFromTracker: Float = 15f,
+) {
+	private val constraintFunction = constraintTypeToFunc(constraintType)
+	private val twistRad = Math.toRadians(twist.toDouble()).toFloat()
+	private val swingRad = Math.toRadians(swing.toDouble()).toFloat()
+	private val allowedDeviationRad = Math.toRadians(allowedDeviation.toDouble()).toFloat()
+	private val maxDeviationFromTrackerRad = Math.toRadians(maxDeviationFromTracker.toDouble()).toFloat()
+	var hasTrackerRotation = false
+
+	/**
+	 * If false solve with minimal movement applied to this link
+	 */
+	var allowModifications = true
+
+	var initialRotation = Quaternion.IDENTITY
+
+	/**
+	 * Apply rotational constraints and if applicable force the rotation
+	 * to be unchanged unless it violates the constraints
+	 */
+	fun applyConstraint(rotation: Quaternion, thisBone: Bone): Quaternion = constraintFunction(rotation, thisBone, swingRad, twistRad, allowedDeviationRad).unit()
```

Comment:
Perhaps we could remove Bone from the scope of these constraints entirely and replace `rotation` with `localRotation: Quaternion` as the current bone's local rotation and `thisBone` could be replaced with `parentRotation: Quaternion` as the parent bone's global rotation? I don't think any of the constraints use anything else besides `COMPLETE`, but that could be handled in another way (or maybe we just take current global rotation instead? it could also be handled earlier in the helper functions, see below).

We could instead use helper functions to accept a global current rotation and a Bone, that way we have minimal input for constraints. This would be beneficial for unit testing and reduce complexity and duplication.

Reply:
I refactored to get very close to what you suggested here but annoyingly bone still needs to be an input to applyConstraint for Complete constraints to work properly in all use cases. This is because when these constraints are being used in an IK solver the input rotation is not necessarily going to be the same rotation as the bone's rotation. In the way it's being used right now the rotation is the same but it's not always going to be the case.

Reply:
Thank you for the additional context, looks good then.